minard/channeldb.py

```diff
@@ -314,38 +315,55 @@ def get_all_thresholds(run):
     maxed thresholds.
     """
 
+    message = ""
+
     conn = engine.connect()
 
     zero = [[0 for i in range(16)] for j in range(19)]
     thr = [[0 for i in range(16)] for j in range(19)]
     disc = [9999]*9728
 
-    # Select most recent zdisc with ecalid field.
-    result = conn.execute("SELECT crate, slot, zero_disc FROM zdisc WHERE "
-        "(ecalid <> '') ORDER BY timestamp DESC limit 304")
-   
-    if result is None:
-        return None
+    if run == 0:
+        result = conn.execute("SELECT crate, slot, vthr FROM current_detector_state "
+            "ORDER BY crate, slot")
+    else:
+        result = conn.execute("SELECT crate, slot, vthr FROM detector_state WHERE "
+            "run = %s ORDER BY crate, slot", run)
 
     rows = result.fetchall()
-    for crate, slot, zero_disc in rows:
-        zero[crate][slot] = zero_disc
-
-    result = conn.execute("SELECT crate, slot, vthr FROM detector_state WHERE "
-        "run = %s ORDER BY crate, slot", run)
 
-    if result is None:
-        return None
+    if not rows:
+        message = "Failure getting vthr information."
+        return 0, 0, 0, message
 
-    rows = result.fetchall()
     for crate, slot, vthr in rows:
         thr[crate][slot] = vthr
 
+    if run == 0:
+        run = detector_state.get_latest_run()
+
+    result = conn.execute("SELECT timestamp from run_state where run = %s", run)
+    time = result.fetchone()
+
+    # Select the ZDISC information with the timestamp before the requested
+    # VTHR information.
+    result = conn.execute("SELECT crate, slot, zero_disc FROM zdisc WHERE "
+        "(ecalid <> '')  and timestamp < %s ORDER BY timestamp DESC limit 304", time)
+
+    rows = result.fetchall()
+
+    # Result is empty, probably zdisc info didn't exist
+    if not rows:
+        message = "Failure getting zdisc info, only goes back to run 103216." 
+        return 0, 0, 0, message
```

Comment:
Does this function normally return lists or single numbers? If lists, might be better to return empty lists instead of 0.